src/Razor/Microsoft.AspNetCore.Razor.Language/test/IntegrationTests/ComponentCodeGenerationTestBase.cs

```diff
@@ -3152,6 +3152,391 @@ public class MyComponent<TItem> : ComponentBase
+        [Parameter] public RenderFragment ChildContent { get; set; }
+    }
+
+    public class Column<TItem> : ComponentBase
+    {
+    }
+}
+"));
+
+            // Act
+            var generated = CompileToCSharp(@"
+<Grid Items=""@(Array.Empty<DateTime>())""><Column /><Column /></Grid>");
+
+            // Assert
+            AssertDocumentNodeMatchesBaseline(generated.CodeDocument);
+            AssertCSharpDocumentMatchesBaseline(generated.CodeDocument);
+            CompileToAssembly(generated);
+        }
+
+        [Fact]
+        public void ChildComponent_Generic_TypeInference_MultiLayerCascaded()
+        {
+            // Arrange
+            AdditionalSyntaxTrees.Add(Parse(@"
+using Microsoft.AspNetCore.Components;
+
+namespace Test
+{
+    [CascadingTypeParameter(nameof(TItem))]
+    public class Ancestor<TItem> : ComponentBase
+    {
+        [Parameter] public System.Collections.Generic.IEnumerable<TItem> Items { get; set; }
+        [Parameter] public RenderFragment ChildContent { get; set; }
+    }
+
+    public class Passthrough : ComponentBase
+    {
+        [Parameter] public RenderFragment ChildContent { get; set; }
+    }
+
+    public class Child<TItem> : ComponentBase
+    {
+    }
+}
+"));
+
+            // Act
+            var generated = CompileToCSharp(@"
+<Ancestor Items=""@(Array.Empty<DateTime>())""><Passthrough><Child /></Passthrough></Ancestor>");
+
+            // Assert
+            AssertDocumentNodeMatchesBaseline(generated.CodeDocument);
+            AssertCSharpDocumentMatchesBaseline(generated.CodeDocument);
+            CompileToAssembly(generated);
+        }
+
+        [Fact]
+        public void ChildComponent_Generic_TypeInference_OverrideCascade()
+        {
+            // This test is to show that, even if an ancestor is trying to cascade its generic types,
+            // a descendant can still override that through inference
+
+            // Arrange
+            AdditionalSyntaxTrees.Add(Parse(@"
+using Microsoft.AspNetCore.Components;
+
+namespace Test
+{
+    [CascadingTypeParameter(nameof(TItem))]
+    public class Grid<TItem> : ComponentBase
+    {
+        [Parameter] public System.Collections.Generic.IEnumerable<TItem> Items { get; set; }
+        [Parameter] public RenderFragment ChildContent { get; set; }
+    }
+
+    public class Column<TItem> : ComponentBase
+    {
+        [Parameter] public TItem OverrideParam { get; set; }
+    }
+}
+"));
+
+            // Act
+            var generated = CompileToCSharp(@"
+<Grid Items=""@(Array.Empty<DateTime>())""><Column OverrideParam=""@(""Some string"")"" /></Grid>");
+
+            // Assert
+            AssertDocumentNodeMatchesBaseline(generated.CodeDocument);
+            AssertCSharpDocumentMatchesBaseline(generated.CodeDocument);
+            CompileToAssembly(generated);
+        }
+
+        [Fact]
+        public void ChildComponent_Generic_TypeInference_NotCascaded()
```

Comment:
How is this test different from `ChildComponent_Generic_Explicit_NotCascaded`

Reply:
ChildComponent_Generic_Explicit_NotCascaded:

```razor
<Grid TItem=""DateTime"" Items=""@(Array.Empty<DateTime>())""><Column /></Grid>
```

ChildComponent_Generic_TypeInference_NotCascaded:

```razor
<Grid Items=""@(Array.Empty<DateTime>())""><Column /></Grid>
```

In the explicit case, the parent has an explicit `TItem=`. It's consequential for the implementation. There's a specific code path that looks for an explicit type parameter on the type parameter source.